app/src/eth_utils.c

```diff
@@ -0,0 +1,216 @@
+/*******************************************************************************
+ *   (c) 2019 Zondax GmbH
+ *   (c) 2022 Zondax GmbH
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ ********************************************************************************/
+#include "eth_utils.h"
+#include <stdio.h>
+#include <zxmacros.h>
+
+#define CHECK_RLP_LEN(BUFF_LEN, RLP_LEN)                                       \
+    {                                                                          \
+        uint64_t buff_len = BUFF_LEN;                                          \
+        uint64_t rlp_len = RLP_LEN;                                            \
+        if (buff_len < rlp_len)                                                \
+            return rlp_no_data;                                                \
+    }
+
+uint64_t
+saturating_add(uint64_t a, uint64_t b)
+{
+
+    uint64_t num = a + b;
+    if (num < a || num < b)
+        return UINT64_MAX;
+
+    return num;
+}
+
+uint32_t
+saturating_add_u32(uint32_t a, uint32_t b)
+{
+
+    uint32_t num = a + b;
+
+    if (num < a || num < b)
+        return UINT32_MAX;
+
+    return num;
+}
+
+int
+be_bytes_to_u64(const uint8_t *bytes, uint8_t len, uint64_t *num)
+{
+    if (bytes == NULL || num == NULL || len == 0 || len > sizeof(uint64_t)) {
+        return -1;
+    }
+
+    *num = 0;
+
+    // fast path
+    if (len == 1) {
+        *num = bytes[0];
+        return 0;
+    }
+
+    uint8_t *num_ptr = (uint8_t *)num;
+    for (int i = len; i--;) {
```

Comment:
This was improved a little bit here: https://github.com/Zondax/ledger-filecoin/blob/main/app/src/eth_utils.c